docarray/utils/reduce.py

```diff
@@ -0,0 +1,164 @@
+    list_fields: List[str] = []
+    set_fields: List[str] = []
+    nested_docs_fields: List[str] = []
+    nested_docarray_fields: List[str] = []
+
+    for field_name, field in doc.__fields__.items():
+        field_type = field.outer_type_
+        if not isinstance(field_type, _GenericAlias) and issubclass(
+            field_type, DocumentArray
+        ):
+            nested_docarray_fields.append(field_name)
+        elif isinstance(field_type, _GenericAlias) and field_type.__origin__ is list:
+            list_fields.append(field_name)
+        elif isinstance(field_type, _GenericAlias) and field_type.__origin__ is set:
+            set_fields.append(field_name)
+        v = getattr(doc, field_name)
+        if v:
+            if isinstance(v, BaseDocument):
+                nested_docs_fields.append(field_name)
+            else:
+                simple_non_empty_fields.append(field_name)
+    return [
+        simple_non_empty_fields,
+        list_fields,
+        set_fields,
+        nested_docarray_fields,
+        nested_docs_fields,
+    ]
+
+
+def reduce_docs(doc1: 'BaseDocument', doc2: 'BaseDocument') -> 'BaseDocument':
+    """
+    Reduces doc1 and doc2 into one Document in-place. Changes are applied to doc1.
+    Reducing 2 Documents consists in setting data properties of the second Document
+    to the first Document if they are empty (priority to the left-most Document)
+    and reducing recursively its nested Documents and DocumentArrays
+    :param doc1: first Document to be reduced. Change is applied in-place
+    :param doc2: second Document to be reduced
+    :return The reduced Document
+    """
+    doc1_fields = _types_analysis(doc1)
+    doc1_simple_non_empty_fields = doc1_fields[0]
+    doc1_list_fields = doc1_fields[1]
+    doc1_set_fields = doc1_fields[2]
+    doc1_nested_docarray_fields = doc1_fields[3]
+    doc1_nested_docs_fields = doc1_fields[4]
+
+    doc2_fields = _types_analysis(doc2)
+    doc2_simple_non_empty_fields = doc2_fields[0]
+    doc2_list_fields = doc2_fields[1]
+    doc2_set_fields = doc2_fields[2]
+    doc2_nested_docarray_fields = doc2_fields[3]
+    doc2_nested_docs_fields = doc2_fields[4]
+
+    # update only fields that are set in doc2 and not set in doc1
+    update_simple_fields = set(doc2_simple_non_empty_fields) - set(
+        doc1_simple_non_empty_fields
+    )
+
+    for field in update_simple_fields:
+        setattr(doc1, field, getattr(doc2, field))
+
+    for field in set(doc1_nested_docs_fields + doc2_nested_docs_fields):
+        setattr(doc1, field, reduce_docs(getattr(doc1, field), getattr(doc2, field)))
+
+    for field in set(doc1_list_fields + doc2_list_fields):
+        array1 = getattr(doc1, field)
+        array2 = getattr(doc2, field)
+        if array1 is None and array2 is not None:
+            setattr(doc1, field, array2)
+        elif array1 is not None and array2 is not None:
+            array1.extend(array2)
+            setattr(doc1, field, array1)
```

Comment:
I think this behaviour (extending the list) is not really expected from the docstring; i would have expected it to be overridden treated like a simple field. It makes sense to do it this way, but it may need an additional sentence to explain